### PR TITLE
Public APIs

### PR DESCRIPTION
**Pull request by Karolis2011**
Makes some APIs Public for _reasons_.

### CODE REVIEW

File: code/datums/api.dm [hunk 1]
```diff
@@ -633,10 +633,50 @@ proc/api_update_command_database()
...
+
+	for(var/client/C in clients)
+		if(C.holder)
+			if(C.holder.fakekey)
+				continue	//so stealthmins aren't revealed by the hub
+			admins++
+		n++
+
+	s["players"] = n
+	s["admins"] = admins
```
Comment (Arrow768) [inline code/datums/api.dm]: Remove the admincount.
For the public its not important to know if there are admins online or not.

**Comment by Arrow768** [comment]
Should also target dev.

File: code/datums/api.dm (hunk 1 above)
Comment (Lohikar) [reply code/datums/api.dm]: grub-mkconfig -o /boot/grub/grub.cfg

Comment (Lohikar) [reply code/datums/api.dm]: This is already exposed via. the public bot, why not have it here?

Comment (HeyBanditoz) [reply code/datums/api.dm]: Disagree. Players should be able to know how many staff are online.

**Comment by Karolis2011** [comment]
Seems like this is not needed or going to be merged.

**Comment by Karolis2011** [comment]
We shouldn't stand out as ones without any public API. So this probably should return.

**Comment by Lohikar** [comment]
It would be nice to have `playerlist` (list of connected non-stealthmin ckeys) and `map` (`current_map.name`) exposed as well.

File: code/datums/api.dm [hunk 2]
```diff
@@ -633,10 +633,48 @@ proc/api_update_command_database()
...
+	s["admins"] = 0
+	s["stationtime"] = worldtime2text()
+	s["roundduration"] = get_round_duration_formatted()
+	s["gameid"] = ""
+	s["map"] = current_map.full_name
+
+	var/n = 0
+
+	for(var/client/C in clients)
+		if(C.holder)
```
Comment (Arrow768) [inline code/datums/api.dm]: This should be changed so ccia/devs arnt counted as admins

File: code/datums/api.dm [hunk 3]
```diff
@@ -633,10 +633,48 @@ proc/api_update_command_database()
...
 		return 1
 
+//Get Server Status
+/datum/topic_command/get_serverstatus_public
+	name = "get_serverstatus_public"
+	description = "Gets the serverstatus"
+	no_auth = 1
+
+/datum/topic_command/get_serverstatus/run_command(queryparams)
+	var/list/s[] = list()
```
Comment (Lohikar) [inline code/datums/api.dm]: Minor nitpick: this should be `var/list/s = list()` or `var/s[]` (preferably former), not a mix of the two.

File: code/datums/api.dm [hunk 4]
```diff
@@ -633,10 +633,48 @@ proc/api_update_command_database()
...
+	s["version"] = game_version
+	s["mode"] = master_mode
+	s["respawn"] = config.abandon_allowed
+	s["enter"] = config.enter_allowed
+	s["vote"] = config.allow_vote_mode
+	s["ai"] = config.allow_ai
+	s["host"] = host ? host : null
+	s["players"] = 0
```
Comment (Lohikar) [inline code/datums/api.dm]: Is it intended that admins is always reported as 0?

File: code/datums/api.dm [hunk 5]
```diff
@@ -633,10 +633,48 @@ proc/api_update_command_database()
...
+	s["admins"] = 0
+	s["stationtime"] = worldtime2text()
+	s["roundduration"] = get_round_duration_formatted()
+	s["gameid"] = ""
```
Comment (Lohikar) [inline code/datums/api.dm]: Is it intended that this is always blank?

File: code/datums/api.dm (hunk 4 above)
Comment (Lohikar) [reply code/datums/api.dm]: Oops, meant to highlight the line below this one.

File: code/datums/api.dm (hunk 5 above)
Comment (Karolis2011) [reply code/datums/api.dm]: Intended to make this information actually not provided. Probably it's other thing if bot's can't handle missing values.

File: code/datums/api.dm [hunk 6]
```diff
@@ -633,12 +633,67 @@ proc/api_update_command_database()
...
+	s["enter"] = config.enter_allowed
+	s["vote"] = config.allow_vote_mode
+	s["ai"] = config.allow_ai
+	s["host"] = host ? host : null
+	s["players"] = 0
+	s["stationtime"] = worldtime2text()
+	s["roundduration"] = get_round_duration_formatted()
+	s["map"] = current_map.full_name
+
+	if(queryparams["status"] == "2")
```
Comment (Arrow768) [inline code/datums/api.dm]: What is the purpose of the status parameter ? - It is not documented

File: code/datums/api.dm [hunk 7]
```diff
@@ -633,12 +633,67 @@ proc/api_update_command_database()
 		data = null
 		return 1
 
+//Get Server Status
+/datum/topic_command/get_serverstatus_public
+	name = "get_serverstatus_public"
+	description = "Gets the serverstatus"
+	no_auth = 1
+
+/datum/topic_command/get_serverstatus/run_command(queryparams)
```
Comment (Arrow768) [inline code/datums/api.dm]: Must be get_serverstatus_public

File: code/datums/api.dm [hunk 8]
```diff
@@ -28,10 +28,10 @@
 		log_debug("API: Throttling bypassed - Command [command.name] set to no_throttle")
 	else
 		if(world_api_rate_limit[addr] != null && config.api_rate_limit_whitelist[addr] == null) //Check if the ip is in the rate limiting list and not in the whitelist
-			if(abs(world_api_rate_limit[addr] - world.time) < config.api_rate_limit) //Check the last request time of the ip
-				world_api_rate_limit[addr] = world.time // Set the time of the last request
+			if(abs(world_api_rate_limit[addr] - world.realtime) < config.api_rate_limit) //Check the last request time of the ip
```
Comment (Lohikar) [inline code/datums/api.dm]: How well does this handle midnight rollover?

Comment (Lohikar) [reply code/datums/api.dm]: Ah, never mind, this is time passed since an epoch, not ToD.

**Comment by skull132** [comment]
!review

File: code/datums/api.dm (hunk 8 above)
Comment (Lohikar) [inline code/datums/api.dm]: As said on discord, `realtime` is not precise enough (~52 second error) for the default value of api_rate_limit (5 seconds). `REALTIMEOFDAY` is likely what you want.

**Comment by Lohikar** [comment]
This still alive?

**Comment by Karolis2011** [comment]
So in summary this:

1. Adds map fields to `get_serverstatus`
2. Makes `get_serverstatus` accessible without auth.
3. Fixes bug when server sleeps and thortling breaksdown.

Now security questions: should code be adapted to make `get_serverstatus` return less data when there is no auth?

**Comment by Lohikar** [comment]
As-is, this exposes the same information as other servers' public APIs.

**Comment by Lohikar** [comment]
It would be nice if this could expose the number of _active_ players too, like Bay's API does.

**Comment by skull132** [comment]
> It would be nice if this could expose the number of _active_ players too, like Bay's API does.

That would basically turn this API end point into an offensive monolith. Which it arguably already is. I'd remove the lists from here and simply keep the numbers if any further alterations are to be done.

But in present state, I am okay with this.

**Comment by Lohikar** [comment]
Been approved for a few days now. This going to be merged?

**Comment by Lohikar** [comment]
If you're removing playerlist from get_serverstatus, please expose `get_player_list` as well.

**Comment by skull132** [comment]
Would be okay with publishing it as long as invisimin is respected. @Arrow768 ?

**Comment by Arrow768** [comment]
I actually have to disagree on publishing the player list.

The question that has not been answered so far is: "why should we?"
The ckey is the unique identifier of a player and I do not believe there is a reason to publish a list of our players for anyone to access.

If we believe that a service has a legitimate usecase to access this data, a API Key can be issued to that service, but just because there might be a service with a legitimate usecase does not mean that we should publish the playerlist for anyone to read.

**Comment by Lohikar** [comment]
- Most other baylike servers expose the playerlist
- BYOND itself exposes the playerlist to the hub
- Doesn't borealis expose it too?

**Comment by Arrow768** [comment]
>     Most other baylike servers expose the playerlist
Just because someone else does it does not mean that we should do it aswell.
(Argumentum ad populum)

>     BYOND itself exposes the playerlist to the hub
That is not our responsibility as it is beyond what we can control.
If we expose it, it is our responsibility and potentially our liability.

>     Doesn't borealis expose it too?
A service that is part of the "aurora environment" and where we believe that it contributes to the community. (And which is only available on selected discord servers)

**Comment by Lohikar** [comment]
What exactly is the argument against exposing the playerlist?

**Comment by Arrow768** [comment]
The question why we should publish it has not been answered sufficiently.

One of the reasons why we should not publish it is for example Article 5.1(c) of the GDPR:
> Personal data shall be adequate, relevant and limited to what is necessary in relation to the purposes for which they are processed

Even if we leave the GDPR aside, I do not see a need to expose the playerlist to the public.

Why is it necessary that we have to provide a public API that exposes the list of players that are currently playing on our server and why can it not be limited to a private API, where API keys are given to services that have a legitimate use for that data ?

**Comment by Lohikar** [comment]
The playerlist (and status in general) is useful in external status server bots like Bot32 or Wrench. Requiring an API key for already-public information seems like pointless bureaucracy / red tape.

**Comment by Arrow768** [comment]
As explained above:
What byond exposes is not our responsibility.
If the information is already published by other sources (such as byond), then feel free to use it from those sources.
("Just because someone else does, doesn't mean that we should" applies here aswell)

But (again) if we publish it, then it becomes our responsibility.

**Comment by HeyBanditoz** [comment]
Almost every gameserver on the *planet* exposes players.

On February 27, 2019 2:45:17 PM MST, Werner <notifications@github.com> wrote:
>As explained above:
>What byond exposes is not our responsibility.
>If the information is already published by other sources (such as
>byond), then feel free to use it from those sources.
>("Just because someone else does, doesn't mean that we should" applies
>here aswell)
>
>But (again) if we publish it, then it becomes our responsibility.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/Aurorastation/Aurora.3/pull/5049#issuecomment-468043212

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by skull132** [comment]
Note regarding BYOND. You can opt out of being visible on player lists published by BYOND. So saying that BYOND publishes player lists is false, since it respects your "Visibility" option and you can then not be published by it.

We also do not have access to said setting, which I suppose makes sense security/privacy wise, so to echo it, we'd have to implement our own mechanic.

**Comment by skull132** [comment]
Superseded by 6127. Closing.

**Pull request closed without merging**